HPCC-18308 eclcmd should pass eclcc options as debug values

afishbeck opened the pull request:
"ecl run <target> <file> -f-<option>"

Allows arbitrary eclcc options to be passed to the eclcc command line.
Those options should also be passed as workunit debug values formatted
as "eclcc-<option>".

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: b0158dcf3f78d368bbb99bcc77b9d4b64f42f5f9
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.263 seconds |


Install hpccsystems-platform-community_6.4.2-rc3.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@ghalliday Please sanity check this

ghalliday commented on ecl/eclcmd/eclcmd_core.cpp:
```diff
@@ -148,6 +148,8 @@ bool doDeploy(EclCmdWithEclTarget &cmd, IClientWsWorkunits *client, const char *
     expandDefintionsAsDebugValues(cmd.definitions, cmd.debugValues);
     if (cmd.debugValues.length())
     {
+        IArrayOf<IEspNamedValue> debugValues;
```
trivial: not used.

ghalliday commented on ecl/eclccserver/eclccserver.cpp:
```diff
@@ -254,7 +254,8 @@ class EclccCompileThread : implements IPooledThread, implements IErrorReporter,
             //Allow eclcc-xx-<n> so that multiple values can be passed through for the same named debug symbol
             const char * start = option + (*option=='-' ? 1 : 6);
-            const char * dash = strrchr(start, '-');     // position of second dash, if present
+            const char * finger = (*start=='-') ? start+1 : start; //support leading double dash
```
I don't understand this change - can you elaborate?

afishbeck replied:
@ghalliday It's slightly confusing, but the key is that when the option is appended to the eclcc command line the stripped '-' is restored.  I was trying not to refactor the code more than I had to as this is a last minute PR.

`eclccCmd.appendf(" -%s=%s", optName.get(), value);`

Some eclcc options actually have two dashes.  "--nostdinc" for example.  So this change preserves the second dash in the option name "-nostdinc" and then both dashes end up on the command line.

ghalliday commented:
@afishbeck I may be misunderstanding, but it looks like an option -f-x is converted to eclcc-x, but would then be matched in eclccserver as optionx.  I would have expected it to be converted to eclcc--x (then the double minus code would make sense.)

afishbeck committed:
HPCC-18308 eclcmd should pass eclcc options as debug values

"ecl run <target> <file> -f-<option>"

Allows arbitrary eclcc options to be passed to the eclcc command line.
Those options should also be passed as workunit debug values formatted
as "eclcc-<option>".

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
@ghalliday Removed unused variable.  Responded to your other comment.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 8e3fb667a346a4b8cef95d4a4472b49c7dcbb1c6
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.332 seconds |


Install hpccsystems-platform-community_6.4.2-rc3.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@richardkchapman looks good.  I'll let you decide which build it goes in.

richardkchapman commented:
I have a concern that --fastparse and --nostdinc should be being applied to the creation of the archive when ecl command calls out to a local eclcc, but don't necessarily apply to the parsing at the remote end - so while it's ok to fix the -f- processing as you have here, we really need --fastparse and --nostdinc to be also handled explicitly by all ecl commands.

I also suspect we may need to push this out to 6.4.4. There is a workaround of creating the archive independently I think.

afishbeck commented:
@richardkchapman I could easily add those as explicit eclcmd options.

My guess would have been that --nostdinc had no effect on the backend since the depencies were already gathered into the archive at the client side?

I can't seem to find --fastparse.

afishbeck commented:
As mentioned in the meeting, --fastparse was a typo, it's --fastsyntax.
